src/event.cc

```diff
@@ -1,325 +1,251 @@
-        return;
-      }
-    }
+  return true;
 }
 
 //
-// C++ callable function to create a JavaScript Event object.
+// initialize oboe_metadata_t structure from a Metabuf buffer
 //
-Napi::Object Event::NewInstance(Napi::Env env) {
-    Napi::EscapableHandleScope scope(env);
+void initializeOboeMd(oboe_metadata_t &md, Napi::Buffer<uint8_t> b) {
+  md.version = XTR_CURRENT_VERSION;
+
+  // copy IDs
+  const uint kTaskIdBase = 1;
+  const uint kOpIdBase = kTaskIdBase + OBOE_MAX_TASK_ID_LEN;
+  for (uint i = 0; i < OBOE_MAX_TASK_ID_LEN; i++) {
+    md.ids.task_id[i] = b[kTaskIdBase + i];
+  }
+  for (uint i = 0; i < OBOE_MAX_OP_ID_LEN; i++) {
+    md.ids.op_id[i] = b[kOpIdBase + i];
+  }
 
-    Napi::Object o = constructor.New({});
+  // initialize lengths
+  md.task_len = OBOE_MAX_TASK_ID_LEN;
+  md.op_len = OBOE_MAX_OP_ID_LEN;
 
-    return scope.Escape(napi_value(o)).ToObject();
+  // and finally flags.
+  md.flags = b[kOpIdBase + OBOE_MAX_OP_ID_LEN];
 }
 
 //
-// C++ callable function to create a JavaScript Event object.
+// add KV pair
 //
-// This signature includes an optional boolean for whether an edge
-// should be set or not. It defaults to true.
+// returns -1 if error.
 //
-Napi::Object Event::NewInstance(Napi::Env env, oboe_metadata_t* omd, bool edge) {
-  Napi::EscapableHandleScope scope(env);
+int addKvPair(oboe_event_t* event, std::string key, Napi::Value value) {
+  int status = -1;
 
-  Napi::Value v0 = Napi::External<oboe_metadata_t>::New(env, omd);
-  Napi::Value v1 = Napi::Boolean::New(env, edge);
+  if (value.IsBoolean()) {
+    status = oboe_event_add_info_bool(event, key.c_str(), value.As<Napi::Boolean>().Value());
+  } else if (value.IsNumber()) {
+    const double v = value.As<Napi::Number>();
 
-  Napi::Object o = constructor.New({v0, v1});
-
-  return scope.Escape(napi_value(o)).ToObject();
-}
+    // it it's within integer range add an integer
+    if (v > -(pow(2, 53) - 1) && v < pow(2, 53) -1) {
+      status = oboe_event_add_info_int64(event, key.c_str(), v);
+    } else {
+      status = oboe_event_add_info_double(event, key.c_str(), v);
+    }
+  } else if (value.IsString()) {
+    std::string s = value.As<Napi::String>();
 
-//
-// JavaScript callable method to format the event as a string.
-//
-// The string representation of an event is the string representation of
-// the metadata part of the event.
-//
-Napi::Value Event::toString(const Napi::CallbackInfo& info) {
-    oboe_metadata_t* md = &this->event.metadata;
-    char buf[Metadata::fmtBufferSize];
-
-    int rc;
-
-    if (info.Length() == 1 && info[0].IsNumber()) {
-      int fmt = info[0].As<Napi::Number>().Int64Value();
-      // default 1 to a human readable form for historical reasons.
-      if (fmt == 1) {
-        fmt = Metadata::fmtHuman;
-      }
-      rc = Metadata::format(md, sizeof(buf), buf, fmt) ? 0 : -1;
+    // make a stab at detecting binary data
+    if (memchr(s.c_str(), '\0', s.length())) {
```

Comment:
it's a relic from the original code; that's why my comment says "make a stab at detecting". i didn't know the original reason or problem so i've just left it the way it was. find a null in the string is a simple yet very fallible solution. but they're probably all fallible without going through a lot of hoops.

do you think it's needed? i can just add a string and require that a buffer is passed for binary data.

i'm glad you asked because i just sort of presumed it made sense.

Reply:
Just saw that the SWIG wrapper code also uses this check: https://github.com/librato/oboe/blob/master/liboboe/oboe.hpp#L391

Like you say it's not a 100% proof way of telling binary from string but I'm fine with leaving it in there as it's obviously been there for a long time without causing issues.